### PR TITLE
[IMP] website_{slides, profile}: improve karma card and misc usability

### PR DESCRIPTION
**Pull request by umgu-odoo**
Currently, there are severlar issues with the karma card, like:
1/ improve karma display
2/ images of the rank and badges are can be changed by user, but
   after saving the changes are discarded

Apart from that, in the course navbar, the hard-coded white
text color is given, due to which in some themes the text
becomes invisible. And in the courses cards below navbar,
the course title is not editable.

This PR improves the following points:

1/ adds a tooltip on the progress bar that shows "2500/10000 xp".
   Displays the left xp needed to complete the next rank and adds a
   tooltip that shows 'Earn 7500 more karma to reach the next rank".
   If there is no next rank, displays total xp without tooltip.
2/ makes the images editable (in karma card as well as in the
   user profile and in '/profile/ranks_badges' page)
3/ makes the rank title editable
4/ removes the hard-coded white text color from the course
   navbar and make the course title editable

taskID-2853264

### CODE REVIEW

File: addons/website_profile/static/src/scss/website_profile.scss [hunk 1]
```diff
@@ -125,6 +125,21 @@ $owprofile-color-bg: mix($body-bg, #efeff4);
 
     .o_pc_overlay {
         @include o-position-absolute(0,0,0,0);
+        // Allow hover effect on svg paths in order to display bootstrap tooltip,
+        // by pushing the absolute div behind it
+        z-index: -1;
```
Comment (jeh-odoo) [inline addons/website_profile/static/src/scss/website_profile.scss:130]: Hum seems like this line broke the mobile view (the rank image disappear), can you check if you can do something ?
Screenshot from 2022-08-11 17-20-17

**Review by jeh-odoo — COMMENTED** [review]
Hello @umgu-odoo 

Thanks for your work, it's globally good.
I've got some suggestions to reduce a bit the diff.
There is also the mobile view that does not display the rank image, I suspect the `z-index: -1`.
Can you check ?

Cheers!

PS: Can you also update the PR title ? :smile:

File: addons/website_profile/static/src/scss/website_profile.scss [hunk 2]
```diff
@@ -125,6 +125,18 @@ $owprofile-color-bg: mix($body-bg, #efeff4);
...
+
+        // Only display single line for the next rank, and display ellipsis if it can not fit in the card
+        .o_pc_text_overflow {
+            @include o-text-overflow(-webkit-box);
+            width: 60%;
+            white-space: normal;
+            -webkit-box-orient: vertical;
+            -webkit-line-clamp: 1;
+        }
```
Comment (jeh-odoo) [inline addons/website_profile/static/src/scss/website_profile.scss:139]: Last small comment and we are good to good after this.
Is there a reason why you use `-webkit` here (is there a specific case you need this)? As I see, I don't think that's mandatory and once set the display will not be the same between chrome and firefox for example.
Does it not work with just this ?
```scss
.o_pc_text_overflow {
    @include o-text-overflow;
    width: 60%;
}
```

Comment (jeh-odoo) [reply addons/website_profile/static/src/scss/website_profile.scss:139]: Hum, then I think that we could do this. What do you think ?
```suggestion
        .o_pc_text_overflow {
            &:not(.o_editable) {
                @include o-text-overflow;
            }
            width: 60%;
        }
```

Comment (umgu-odoo) [reply addons/website_profile/static/src/scss/website_profile.scss]: Changes are done but i have one question,
If we write a long sentence, it looks like 
https://www.awesomescreenshot.com/image/31638037?key=d636192a370d53204e58c0611fe31868

Will it be okay? what do you think?  

cc: @LNAodoo

File: addons/website_profile/static/src/js/website_profile.js [hunk 3]
```diff
@@ -116,6 +116,19 @@ publicWidget.registry.websiteProfileEditor = publicWidget.Widget.extend({
...
 });
 
+publicWidget.registry.websiteProfileNextRankCard = publicWidget.Widget.extend({
+    selector: '.o_wprofile_progress_circle',
+
+    /**
+     * @override
+     */
+    start: function () {
+        this.$('g[data-bs-toggle="tooltip"]').tooltip({delay: 0});
```
Comment (awa-odoo) [inline addons/website_profile/static/src/js/website_profile.js:126]: Can you try to instead make this tooltip using only bootstrap classes?

Apparently it's possible to pass the delay as data in the template:

https://getbootstrap.com/docs/4.0/components/tooltips/#options

This will allow to get rid of this JS override (and also we tend to avoid jQuery as much as possible now that OWL is merged).

File: addons/website_profile/static/src/scss/website_profile.scss [hunk 4]
```diff
@@ -125,6 +125,18 @@ $owprofile-color-bg: mix($body-bg, #efeff4);
 
     .o_pc_overlay {
         @include o-position-absolute(0,0,0,0);
+        // Allow hover effect on svg paths in order to display bootstrap tooltip,
+        // by pushing the absolute div behind it
+        z-index: -1;
+
+        // Only display single line for the next rank, and display ellipsis if it can not fit in the card
+        .o_pc_text_overflow {
+            @include o-text-overflow(-webkit-box);
```
Comment (awa-odoo) [inline addons/website_profile/static/src/scss/website_profile.scss:134]: Did not see a single occurrence of `@include o-text-overflow(-webkit-box);` so it makes me a bit suspicious.

Is the vendor prefix required here?

We tend to avoid those.

File: addons/website_profile/views/website_profile.xml [hunk 5]
```diff
@@ -315,8 +315,11 @@
                 <t t-else="">
                     <t t-set="user_points" t-value="0"/>
                 </t>
-                <path class="o_pc_circle_bg" d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831" />
-                <path class="o_pc_circle" t-attf-stroke-dasharray="#{user_points}, 100" d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831" stroke="url(#gradient)" mask="url(#mask)"/>
+                <g data-bs-toggle="tooltip"
+                    t-attf-title="{{ ('{0}/{1} xp'.format(user.karma, next_rank_id.karma_min)) if next_rank_id else ''}}">
```
Comment (awa-odoo) [inline addons/website_profile/views/website_profile.xml:319]: Maybe using f-strings would be more readable here.

File: addons/website_profile/views/website_profile.xml [hunk 6]
```diff
@@ -326,25 +329,16 @@
...
-                        t-options="{'format_decimalized_number': True, 'precision_digits': 2}"/>
-                    <span t-else="" class="fw-bold"
-                        t-field="user.rank_id.karma_min"
-                        t-options="{'format_decimalized_number': True, 'precision_digits': 2}"/>
-                     xp
+                <t t-set="rank_id" t-value="next_rank_id if next_rank_id else user.rank_id"/>
+                <div t-field="rank_id.image_1920"
+                    t-attf-style="max-width: {{img_max_width or '50%'}};"
+                    t-options="{'widget': 'image', 'preview_image': 'image_128'}"/>
+                <h4 class="o_pc_text_overflow text-center" t-field="rank_id.name"/>
```
Comment (awa-odoo) [inline addons/website_profile/views/website_profile.xml:336]: Actually, what is the added value compared to a regular "text-overflow" class?
Seems a bit overkill to define all this style just for a label, not done like this everywhere else.

File: addons/website_profile/views/website_profile.xml [hunk 7]
```diff
@@ -326,25 +329,16 @@
...
+                <small class="w-50 text-center">
+                    <t t-set="karma_value" t-value="next_rank_id.karma_min - user.karma if next_rank_id else user.karma"/>
+                    <span t-if="next_rank_id">Get</span>
```
Comment (awa-odoo) [inline addons/website_profile/views/website_profile.xml:339]: Would split this into 2 cases
t-if="next_rank_id"
...
t-else
...

It's currently a bit hard to follow.

File: addons/website_profile/views/website_profile.xml [hunk 8]
```diff
@@ -363,11 +357,8 @@
                                         t-attf-src="/website_profile/static/src/img/badge_#{badge.badge_id.level}.svg"
                                         class="o_object_fit_cover me-0"
                                         t-att-alt="badge.badge_id.name"/>
-                                    <img t-else=""
-                                        width="38" height="38"
-                                        t-att-src="website.image_url(badge.badge_id, 'image_128')"
-                                        class="o_object_fit_cover me-0"
-                                        t-att-alt="badge.badge_id.name"/>
+                                    <div t-else="" t-field="badge.badge_id.image_1920"
```
Comment (awa-odoo) [inline addons/website_profile/views/website_profile.xml:360]: Could we switch those around?

Always better to see "I have an image, show it, otherwise use this fallback" than the other way around.

File: addons/website_slides/views/website_slides_templates_homepage.xml [hunk 9]
```diff
@@ -495,11 +495,9 @@
         <t t-foreach="challenges" t-as="challenge">
             <t t-set="challenge_done" t-value="challenge in challenges_done if challenges_done else False"/>
             <div t-attf-class="d-flex mb-3 align-items-center #{'o_wslides_entry_muted' if not challenge_done else ''}">
-                <img class="me-2"
-                    style="max-height: 36px;"
-                    t-att-src="website.image_url(challenge.reward_id, 'image_128') if challenge.reward_id.image_1920 else
-                               '/website_profile/static/src/img/badge_%s.svg' % (challenge.reward_id.level)"
-                    t-att-alt="challenge.reward_id.name"/>
+                <img t-if="not challenge.reward_id.image_1920" t-attf-src="'/website_profile/static/src/img/badge_%s.svg' % (challenge.reward_id.level)" t-att-alt="challenge.reward_id.name" style="max-height: 36px" class="me-2"/>
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_homepage.xml:498]: Same idea here, would switch those around, show an image if available, otherwise use the fallback.

**Review by awa-odoo — COMMENTED** [review]
Hello @umgu-odoo 

Looking good overall, nice improvement of the website editor compatibility.

I left a few comments to try to polish a bit more.

Cheers!

Side-note: Looks like there is a small file conflict.

File: addons/website_profile/static/src/js/website_profile.js (hunk 3 above)
Comment (dja-odoo) [reply addons/website_profile/static/src/js/website_profile.js]: Moved `delay` to xml, but I don't think currently we can initialize toolitp without jQuery (I also did not obverse any example without jQuery).

File: addons/website_profile/static/src/scss/website_profile.scss (hunk 4 above)
Comment (dja-odoo) [reply addons/website_profile/static/src/scss/website_profile.scss:134]: See https://github.com/odoo/odoo/pull/92871/files#r949843249 to see from where the reference was taken and why. Basically removing vendor prefix breaks edition.

File: addons/website_profile/views/website_profile.xml (hunk 5 above)
Comment (dja-odoo) [reply addons/website_profile/views/website_profile.xml:319]: I tried but did not work. Could you provide an example?

File: addons/website_profile/views/website_profile.xml (hunk 6 above)
Comment (dja-odoo) [reply addons/website_profile/views/website_profile.xml:336]: Same as above, it breaks edition.

File: addons/website_profile/static/src/scss/website_profile.scss (hunk 4 above)
Comment (awa-odoo) [reply addons/website_profile/static/src/scss/website_profile.scss:134]: @dja-odoo 

Just checked out locally and gave it a try, something like this seems to work:

```scss
        h4 {
            width: 60%;

            // only apply ellipsis when in non-edit mode, that way the editor can see the full value
            &:not(.o_editable) {
                overflow: hidden;
                text-overflow: ellipsis;
                white-space: nowrap;
            }
        }
```

For both edition and non-edition mode.
I would really like to avoid vendor prefixes if possible and use standard stuff.
This is just some text we want to crop, there really should not be a need to introduce new rules that are not used anywhere else.

File: addons/website_profile/views/website_profile.xml (hunk 5 above)
Comment (awa-odoo) [reply addons/website_profile/views/website_profile.xml:319]: Maybe it's not supported in XML.

In that case how about a regular:
`{{ '%s/%s xp' % (user.karma, next_rank_id.karma_min) if next_rank_id else ''}}`

?

File: addons/website_profile/views/website_profile.xml (hunk 6 above)
Comment (awa-odoo) [reply addons/website_profile/views/website_profile.xml:336]: (See above suggestion)

File: addons/website_profile/static/src/js/website_profile.js [hunk 10]
```diff
@@ -116,6 +116,19 @@ publicWidget.registry.websiteProfileEditor = publicWidget.Widget.extend({
...
 });
 
+publicWidget.registry.websiteProfileNextRankCard = publicWidget.Widget.extend({
+    selector: '.o_wprofile_progress_circle',
+
+    /**
+     * @override
+     */
+    start: function () {
+        this.$('g[data-bs-toggle="tooltip"]').tooltip();
```
Comment (awa-odoo) [inline addons/website_profile/static/src/js/website_profile.js:126]: Well the doc says:

> Tooltips are opt-in for performance reasons, so you must initialize them yourself.

So you are right, we probably have to have this code.
A bit sad to implement a full publicWidget for a tooltip, but oh well ^^

**Review by awa-odoo — COMMENTED** [review]
Hello @dja-odoo 

Almost there, but I would really like to simplify the text-break thing.

Cheers!

File: addons/website_profile/static/src/scss/website_profile.scss (hunk 4 above)
Comment (dja-odoo) [reply addons/website_profile/static/src/scss/website_profile.scss:134]: Hello @awa-odoo, in edition mode it will look like this for the long string
Screenshot 2022-09-27 at 15 49 38
If that's okay, we can indeed avoid vendor prefix :)

File: addons/website_profile/views/website_profile.xml (hunk 5 above)
Comment (dja-odoo) [reply addons/website_profile/views/website_profile.xml]: Done :)

File: addons/website_profile/static/src/scss/website_profile.scss (hunk 4 above)
Comment (awa-odoo) [reply addons/website_profile/static/src/scss/website_profile.scss:134]: That's fine, it's a very side case, I'd rather not make the code too complex :)

Comment (dja-odoo) [reply addons/website_profile/static/src/scss/website_profile.scss]: @awa-odoo Thanks for the quick reply, done :)

**Comment by awa-odoo** [comment]
@robodoo r+
